File: app/src/main/java/vandyke/siamobile/dialogs/WalletCreateDialog.java
package vandyke.siamobile.dialogs;

import android.app.Dialog;
import android.app.DialogFragment;
import android.app.FragmentManager;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import org.json.JSONException;
import org.json.JSONObject;
import vandyke.siamobile.MainActivity;
import vandyke.siamobile.R;
import vandyke.siamobile.api.SiaRequest;
import vandyke.siamobile.api.Wallet;

public class WalletCreateDialog extends DialogFragment {
    
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = MainActivity.getDialogBuilder(getActivity());
        final View view = getActivity().getLayoutInflater().inflate(R.layout.dialog_wallet_create, null);

        final CheckBox createFromSeed = (CheckBox) view.findViewById(R.id.walletCreateFromSeed);
        final EditText seedField = (EditText) view.findViewById(R.id.walletCreateSeed);
        seedField.setVisibility(View.GONE);
        createFromSeed.setOnClickListener(new View.OnClickListener() {
            public void onClick(View view) {
                if (createFromSeed.isChecked())
                    seedField.setVisibility(View.VISIBLE);
                else
                    seedField.setVisibility(View.GONE);
            }
        });

        final CheckBox forceCheck = (CheckBox) view.findViewById(R.id.walletCreateForce);
        final TextView forceWarning = (TextView) view.findViewById(R.id.walletCreateForceWarning);
        forceWarning.setVisibility(View.GONE);
        forceCheck.setOnClickListener(new View.OnClickListener() {
            public void onClick(View view) {
                if (forceCheck.isChecked())
                    forceWarning.setVisibility(View.VISIBLE);
                else
                    forceWarning.setVisibility(View.GONE);
            }
        });

        builder.setTitle("Create Wallet")
                .setView(view)
                .setPositiveButton("Create", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        String password = ((EditText) view.findViewById(R.id.newPasswordCreate)).getText().toString();
                        if (!password.equals(((EditText) view.findViewById(R.id.confirmNewPasswordCreate)).getText().toString())) {
                            Toast.makeText(getActivity(), "New passwords don't match", Toast.LENGTH_SHORT).show();
                            return;
                        }
                        boolean force = forceCheck.isChecked();
                        String dictionary = "english";
                        if (createFromSeed.isChecked())
                            Wallet.initSeed(password, force, dictionary, seedField.getText().toString(), new SiaRequest.VolleyCallback(getActivity()) {
                                public void onError(SiaRequest.Error error) {
                                    if (error.getReason() == SiaRequest.Error.Reason.WALLET_NOT_ENCRYPTED)
                                        Toast.makeText(getActivity(), "Success. Scanning blockchain, please wait", Toast.LENGTH_SHORT).show();
                                    else
                                        super.onError(error);
                                }
                            });
                        else
                            Wallet.init(password, force, dictionary, new SiaRequest.VolleyCallback(getActivity()));
                    }
                })
                .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {

                    }
                });
        return builder.create();
    }

    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.dialog_wallet_create, null);
    }

    public void onActivityCreated(Bundle bundle) {
        super.onActivityCreated(bundle);
        getDialog().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
    }

    public static void createAndShow(FragmentManager fragmentManager) {
        new WalletCreateDialog().show(fragmentManager, "wallet init dialog");
    }
}

File: app/src/main/java/vandyke/siamobile/ui/wallet/dialogs/WalletCreateDialog.kt
/*
 * Copyright (c) 2017 Nicholas van Dyke
 *
 * This file is subject to the terms and conditions defined in 'LICENSE.md'
 */

package vandyke.siamobile.ui.wallet.dialogs

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.view.View
import android.widget.Toast
import kotlinx.android.synthetic.main.fragment_wallet_create.*
import vandyke.siamobile.R
import vandyke.siamobile.backend.networking.SiaCallback
import vandyke.siamobile.backend.networking.SiaError
import vandyke.siamobile.backend.networking.Wallet
import vandyke.siamobile.backend.wallet.WalletService
import vandyke.siamobile.prefs
import vandyke.siamobile.util.GenUtil
import vandyke.siamobile.util.SnackbarUtil

class WalletCreateDialog : BaseDialogFragment() {
    override val layout: Int = R.layout.fragment_wallet_create

    override fun create(view: View?, savedInstanceState: Bundle?) {
        setCloseListener(walletCreateCancel)
        walletCreateSeed.visibility = View.GONE
        walletCreateFromSeed.setOnClickListener {
            if (walletCreateFromSeed.isChecked)
                walletCreateSeed.visibility = View.VISIBLE
            else
                walletCreateSeed.visibility = View.GONE
        }

        walletCreateForceWarning.visibility = View.GONE
        walletCreateForce.setOnClickListener {
            if (walletCreateForce.isChecked)
                walletCreateForceWarning.visibility = View.VISIBLE
            else
                walletCreateForceWarning.visibility = View.GONE
        }

        walletCreateButton.setOnClickListener(View.OnClickListener {
            val password = newPasswordCreate.text.toString()
            if (password != confirmNewPasswordCreate.text.toString()) {
                SnackbarUtil.snackbar(view, "New passwords don't match", Snackbar.LENGTH_SHORT)
                return@OnClickListener
            }
            val force = walletCreateForce.isChecked
            val dictionary = "english"
            if (!walletCreateFromSeed.isChecked) {
                Wallet.init(password, dictionary, force, SiaCallback({ it ->
                    SnackbarUtil.successSnackbar(view)
                    close()
                    WalletService.singleAction(activity, { it.refresh() })
                    if (prefs.operationMode == "cold_storage")
                        showCsWarning()
                    showSeed(it.primaryseed)
                }, {
                    if (it.reason == SiaError.Reason.WALLET_SCAN_IN_PROGRESS) {
                        SnackbarUtil.snackbar(view, "Success. Scanning the blockchain for coins belonging to the given seed. Please wait", Snackbar.LENGTH_LONG)
                        close()
                    } else {
                        it.snackbar(view)
                    }
                }))
            } else {
                Wallet.initSeed(password, dictionary, walletCreateSeed.text.toString(), force, SiaCallback({ ->
                    SnackbarUtil.successSnackbar(view)
                    close()
                    WalletService.singleAction(activity, { it.refresh() })
                    if (prefs.operationMode == "cold_storage")
                        showCsWarning()
                }, {
                    if (it.reason == SiaError.Reason.WALLET_SCAN_IN_PROGRESS) {
                        SnackbarUtil.snackbar(view, "Success. Scanning the blockchain for coins belonging to the given seed. Please wait", Snackbar.LENGTH_LONG)
                        close()
                    } else {
                        it.snackbar(view)
                    }
                }))
            }
        })
    }

    private fun showCsWarning() {
        GenUtil.getDialogBuilder(activity)
                .setTitle("IMPORTANT")
                .setMessage("You just created a wallet while in cold storage mode. While in cold storage mode," +
                        " Sia Mobile is not connected to the Sia network and does not have a copy of the Sia blockchain. Normally this would mean you can't " +
                        "view your balance and transactions, but Sia Mobile ESTIMATES your CONFIRMED balance and transactions once per minute using explore.sia.tech." +
                        "\n\nIt also means certain functions that require the blockchain or a connection to the network will" +
                        " be unavailable - most importantly, you can't send coins from a cold storage wallet. If you wish to use these unavailable functions," +
                        " you can, AT ANY TIME, load your wallet seed on a full node (such as Sia-UI for desktop, or Sia Mobile in local full node mode)" +
                        " to access and interact with your previously received coins.")
                .setPositiveButton("I have read and understood this", null)
                .show()
    }

    private fun showSeed(seed: String) {
        val msg = "Below is your wallet seed. Your wallet's addresses are generated using this seed. Therefore, any coins you " +
                "send to this wallet and its addresses will \"belong\" to this seed. It's what you will need" +
                " in order to recover your coins if something happens to your wallet, or to load your wallet on another device. Keep it safe."
        GenUtil.getDialogBuilder(activity)
                .setTitle("Wallet seed")
                .setMessage("$msg\n\n$seed")
                .setPositiveButton("Copy seed", { _, _ ->
                    GenUtil.copyToClipboard(activity, seed)
                    Toast.makeText(activity, "Copied seed", Toast.LENGTH_SHORT).show()
                })
                .setNegativeButton("Close", null)
                .show()
    }
}
